crates/bevy_ecs/src/storage/sparse_set.rs

```diff
@@ -0,0 +1,508 @@
+use crate::{
+    component::{ComponentFlags, ComponentId, ComponentInfo},
+    entity::Entity,
+    storage::BlobVec,
+};
+use std::{cell::UnsafeCell, marker::PhantomData};
+
+#[derive(Debug)]
+pub struct SparseArray<I, V = I> {
+    values: Vec<Option<V>>,
+    marker: PhantomData<I>,
+}
+
+impl<I: SparseSetIndex, V> Default for SparseArray<I, V> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<I, V> SparseArray<I, V> {
+    #[inline]
+    pub const fn new() -> Self {
+        Self {
+            values: Vec::new(),
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<I: SparseSetIndex, V> SparseArray<I, V> {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            values: Vec::with_capacity(capacity),
+            marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub fn insert(&mut self, index: I, value: V) {
+        let index = index.sparse_set_index();
+        if index >= self.values.len() {
+            self.values.resize_with(index + 1, || None);
+        }
+        self.values[index] = Some(value);
+    }
+
+    #[inline]
+    pub fn contains(&self, index: I) -> bool {
+        let index = index.sparse_set_index();
+        self.values.get(index).map(|v| v.is_some()).unwrap_or(false)
+    }
+
+    #[inline]
+    pub fn get(&self, index: I) -> Option<&V> {
+        let index = index.sparse_set_index();
+        self.values.get(index).map(|v| v.as_ref()).unwrap_or(None)
+    }
+
+    #[inline]
+    pub fn get_mut(&mut self, index: I) -> Option<&mut V> {
+        let index = index.sparse_set_index();
+        self.values
+            .get_mut(index)
+            .map(|v| v.as_mut())
+            .unwrap_or(None)
+    }
+
+    #[inline]
+    pub fn remove(&mut self, index: I) -> Option<V> {
+        let index = index.sparse_set_index();
+        self.values.get_mut(index).and_then(|value| value.take())
+    }
+
+    #[inline]
+    pub fn get_or_insert_with(&mut self, index: I, func: impl FnOnce() -> V) -> &mut V {
+        let index = index.sparse_set_index();
+        if index < self.values.len() {
+            // SAFE: just checked bounds
+            let value = unsafe { self.values.get_unchecked_mut(index) };
+            if value.is_none() {
+                *value = Some(func());
+            }
+
+            return value.as_mut().unwrap();
+        }
```

Comment:
Would the following work?
```suggestion
        if let Some(mut value) = self.values.get_mut(index) {
            if value.is_none() {
                *value = Some(func());
            }

            return value.as_mut().unwrap();
        }
```

Reply:
Sadly no. The current code was a reaction to the fact that `if let Some(value) = self.values.get_mut(index)` holds on to the mutable borrow:

![image](https://user-images.githubusercontent.com/2694663/109400369-3bc8b380-78fd-11eb-9d8f-b28525abc4b2.png)

Reply:
```suggestion
     match self.values.get_mut(index) {
            Some(mut value) => return value.get_or_insert_with(func),
            None=>()
         }
```
could work?

Reply:
Same problem:

![image](https://user-images.githubusercontent.com/2694663/109401030-1178f500-7901-11eb-95cc-d77d2cbf4c07.png)

It doesn't help reduce unsafe, but we might as well use get_or_insert_with to make the impl simpler:

![image](https://user-images.githubusercontent.com/2694663/109401047-366d6800-7901-11eb-8cdf-2e19010fddb7.png)

Reply:
Haha the second screenshot above has an error (forgot a return). Causes on ICE on nightly: https://github.com/rust-lang/rust/issues/82612.

Reply:
This should be fixed with polonius I think. By the way I think LLVM would optimize away the bound check if you used `self.values[index]` inside the if.

Reply:
> This should be fixed with polonius I think
yeah that was my understanding as well

> By the way I think LLVM would optimize away the bound check if you used self.values[index] inside the if.
ill give it a whirl to see if it affects benchmarks

Reply:
@bjorn3 switching to self.values[index] did not regress perf. good call!